internal/controller/runtime/api_discovery.go

```diff
@@ -0,0 +1,246 @@
+	for _, want := range r.want {
+		if wantAPIs[want.Group] == nil {
+			wantAPIs[want.Group] = make(map[string]sets.Set[string])
+		}
+		if wantAPIs[want.Group][want.Version] == nil {
+			wantAPIs[want.Group][want.Version] = sets.New[string]()
+		}
+		if want.Kind != "" {
+			wantAPIs[want.Group][want.Version].Insert(want.Kind)
+		}
+	}
+
+	// Fetch Groups and Versions from Kubernetes.
+	groups, err := r.Client.ServerGroups()
+	if err != nil {
+		return err
+	}
+
+	// Build an index of the Groups, GVs, GKs, and GVKs available in Kuberentes
+	// that we want to know about.
+	haveWantedAPIs := make(map[API]struct{})
+	for _, apiG := range groups.Groups {
+		var haveG string = apiG.Name
+		haveWantedAPIs[API{Group: haveG}] = struct{}{}
+
+		for _, apiGV := range apiG.Versions {
+			var haveV string = apiGV.Version
+			haveWantedAPIs[API{Group: haveG, Version: haveV}] = struct{}{}
+
+			// Only fetch Resources when there are Kinds we want to know about.
+			if wantAPIs[haveG][""].Len() == 0 && wantAPIs[haveG][haveV].Len() == 0 {
+				continue
+			}
+
+			resources, err := r.Client.ServerResourcesForGroupVersion(apiGV.GroupVersion)
+			if err != nil {
+				return err
+			}
+
+			for _, apiR := range resources.APIResources {
+				var haveK string = apiR.Kind
+				haveWantedAPIs[API{Group: haveG, Kind: haveK}] = struct{}{}
+				haveWantedAPIs[API{Group: haveG, Kind: haveK, Version: haveV}] = struct{}{}
+			}
+		}
+	}
+
+	r.have.Lock()
+	r.have.APISet = haveWantedAPIs
+	r.have.Unlock()
+
+	return nil
+}
+
+// Start periodically reads the Kuberentes API. It blocks until ctx is cancelled.
+func (r *APIDiscoveryRunner) Start(ctx context.Context) error {
+	ticker := time.NewTicker(r.refresh)
+	defer ticker.Stop()
+
+	log := logging.FromContext(ctx).WithValues("controller", "kubernetes")
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := r.Read(); err != nil {
+				log.Error(err, "Unable to detect Kubernetes APIs")
+			}
+		case <-ctx.Done():
+			// TODO(controller-runtime): Fixed in v0.19.0
+			// https://github.com/kubernetes-sigs/controller-runtime/issues/1927
+			if errors.Is(ctx.Err(), context.Canceled) {
+				return nil
+			}
+			return ctx.Err()
+		}
+	}
+}
+
+// Has returns true when api is available in Kuberentes.
+func (r *APIDiscoveryRunner) Has(api API) bool { return r.HasOne(api) }
+
+// HasAll returns true when every api is available in Kubernetes.
+func (r *APIDiscoveryRunner) HasAll(api ...API) bool {
+	r.have.RLock()
+	defer r.have.RUnlock()
+	return r.have.HasAll(api...)
+}
+
+// HasOne returns true when at least one api is available in Kubernetes.
+func (r *APIDiscoveryRunner) HasOne(api ...API) bool {
+	r.have.RLock()
+	defer r.have.RUnlock()
+	return r.have.HasOne(api...)
+}
+
+type apiContextKey struct{}
+
+// Kubernetes returns the APIs previously stored by [NewAPIContext].
+// When nothing was stored, it returns an empty [APISet].
+func Kubernetes(ctx context.Context) APIs {
```

Comment:
❓ What do you think of this interface/experience? IIUC, callers will look like this:

```go
if runtime.Kubernetes(ctx).Has(runtime.API{
  Group: "snapshot.storage.k8s.io", Kind: "VolumeSnapshot",
}) {
  // ...
}
```

Reply:
Yeah this feels good.